File: src/main/java/opennlp/textgrounder/util/MathUtil.java
///////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Jason Baldridge, The University of Texas at Austin
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
///////////////////////////////////////////////////////////////////////////////
package opennlp.textgrounder.util;

import java.util.Arrays;

/**
 * Handy functions for common mathematical operations on data
 * structures, plus computations in log-space.
 *
 * @author  Jason Baldridge
 * @version $Revision: 1.53 $, $Date: 2006/10/12 21:20:44 $
 */
public class MathUtil {

    public static double LOG_ZERO = Double.NEGATIVE_INFINITY;

    public static double eexp (double x) { 
	if (Double.isInfinite(x))
	    return 0.0;
	else
	    return Math.exp(x);
    }

    public static double elog (double x) { 
	if (x == 0.0)
	    return LOG_ZERO;
	else
	    return Math.log(x);
    }

    public static double elogProduct(double ... values) {
	double sum = 0.0;
	for (double x: values) {
	    if (Double.isInfinite(x))
		return LOG_ZERO;
	    else
		sum += x;
	}
	return sum;
    }

    public static double elogSum(double x, double y) {
	if (Double.isInfinite(x))
	    return y;
	else if (Double.isInfinite(y))
	    return x;
	else {
	    if (x > y)
		return x + Math.log1p(Math.exp(y-x));
//		return x + Math.log(1 + Math.exp(y-x));
	    else
		return x + Math.log1p(Math.exp(x-y));
//		return y + Math.log(1 + Math.exp(x-y));
	}
    }

    public static void logNormalize(double[] vals) {
	double total = sum(vals);
	for (int i=0; i<vals.length; i++)
	    vals[i] = elog(vals[i]/total);
    }

    public static double sum (double[] values) {
	double total = 0.0;
	for (int i=0; i<values.length; i++)
	    total += values[i];
	return total;
    }

    public static double entropy(double[] vals) {
	double entropy = 0.0;
	for (int i=0; i<vals.length; i++)
	    entropy -= vals[i]*Math.log(vals[i]);
	return entropy;
    }

    public static double entropyOfLogDistribution(double[] vals) {
	double entropy = 0.0;
	for (int i=0; i<vals.length; i++)
	    entropy -= vals[i]*Math.exp(vals[i]);
	return entropy;
    }

    public static final int getIndexOfHighest (double[] vals) {
	int highestIndex = 0;
	double highestVal = vals[0];
	for (int i=1; i<vals.length; i++) {
	    if (vals[i] > highestVal) {
		highestVal = vals[i];
		highestIndex = i;
	    }
	}
	return highestIndex;
    }

    public static final int[] rankScores (double[] vals) {
	IntDoublePair[] paired = new IntDoublePair[vals.length];
	for (int i=0; i<vals.length; i++)
	    paired[i] = new IntDoublePair(i, vals[i]);

	Arrays.sort(paired);

	int[] ranked = new int[vals.length];
	for (int i=0; i<vals.length; i++)
	    ranked[i] = paired[i].intValue;

	return ranked;
    }

    public static final void exponentiate (double[] vals) {
	for (int i=0; i<vals.length; i++)
	    vals[i] = Math.exp(vals[i]);
    }


    public static final void takeLogarithm (double[] vals) {
	for (int i=0; i<vals.length; i++)
	    vals[i] = Math.log(vals[i]);
    }

    public static final double[] uniformLogDistribution (int numValues) {
	double[] probs = new double[numValues];
	Arrays.fill(probs, Math.log(1.0/numValues));
	return probs;
    }

    public static final void normalize (double[] vals) {
	double total = 0.0;
	for (int i=0; i<vals.length; i++)
	    total += vals[i];
	for (int i=0; i<vals.length; i++)
	    vals[i] = vals[i]/total;
    }

    public static final String printProbabilities (double[] vals) {
	StringBuffer sb = new StringBuffer();
	sb.append('[');
	sb.append(Constants.PERCENT_FORMAT.format(vals[0]));
	for (int i=1; i<vals.length; i++)
	    sb.append(',').append(Constants.PERCENT_FORMAT.format(vals[i]));
	sb.append(']');
	return sb.toString();
    }


}

File: src/main/scala/opennlp/textgrounder/util/mathutil.scala
///////////////////////////////////////////////////////////////////////////////
//  mathutil.scala
//
//  Copyright (C) 2011, 2012 Ben Wing, The University of Texas at Austin
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
///////////////////////////////////////////////////////////////////////////////

package opennlp.textgrounder.util

import math._

package object mathutil {
  /**
   *  Return the median value of a list.  List will be sorted, so this is O(n).
   */
  def median(list: Seq[Double]) = {
    val sorted = list.sorted
    val len = sorted.length
    if (len % 2 == 1)
      sorted(len / 2)
    else {
      val midp = len / 2
      0.5*(sorted(midp-1) + sorted(midp))
    }
  }
  
  /**
   *  Return the mean of a list.
   */
  def mean(list: Seq[Double]) = {
    list.sum / list.length
  }

  def variance(x: Seq[Double]) = {
    val m = mean(x)
    mean(for (y <- x) yield ((y - m) * (y - m)))
  }

  def stddev(x: Seq[Double]) = sqrt(variance(x))

  abstract class MeanShift[Coord : Manifest](
      h: Double = 1.0,
      max_stddev: Double = 1e-10,
      max_iterations: Int = 100
    ) {
    def squared_distance(x:Coord, y:Coord): Double
    def weighted_sum(weights:Array[Double], points:Array[Coord]): Coord
    def scaled_sum(scalar:Double, points:Array[Coord]): Coord

    def vec_mean(points:Array[Coord]) = scaled_sum(1.0/points.length, points)

    def vec_variance(points:Array[Coord]) = {
      def m = vec_mean(points)
      mean(
        for (i <- 0 until points.length) yield squared_distance(m, points(i)))
    }

    def mean_shift(list: Seq[Coord]):Array[Coord] = {
      var next_stddev = max_stddev + 1
      var numiters = 0
      val points = list.toArray
      val shifted = list.toArray
      while (next_stddev >= max_stddev && numiters <= max_iterations) {
        for (j <- 0 until points.length) {
          val y = shifted(j)
          val weights =
            (for (i <- 0 until points.length)
             yield exp(-squared_distance(y, points(i))/(h*h)))
          val weight_sum = weights sum
          val normalized_weights = weights.map(_ / weight_sum).toArray
          shifted(j) = weighted_sum(normalized_weights, points)
        }
        numiters += 1
        next_stddev = sqrt(vec_variance(shifted))
      }
      shifted
    }
  }
} 
